packages/blocks/src/api/registration.js

```diff
@@ -232,6 +232,27 @@ export function registerBlockType( blockNameOrMetadata, settings ) {
 		return;
 	}
 
+	// Parent being an array hasn't been enforced in the past,
+	// so this is a way to maintain backwards compatibility
+	// with 3rd party blocks that may have been using it as a string.
+	if (
```

Comment:
Isn't the `parent` prop in metadata (if an object is passed)?  If that's the case and the metadata are not passed we should get them and make the check.

Additionally can we simplify the checks here to explicitly check what we want? Something like:

```
if (
	blockNameOrMetadata.parent &&
	! Array.isArray( blockNameOrMetadata.parent ) &&
	typeof blockNameOrMetadata.parent === 'string'
) {
	blockNameOrMetadata.parent = [ blockNameOrMetadata.parent ];
}
```

Reply:
> Isn't the parent prop in metadata (if an object is passed)? If that's the case and the metadata are not passed we should get them and make the check.

Oh, I missed that, thanks for bringing it up! I'm a bit confused about this for some reasons (existing code, [docs](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-registration/) and other [implementations](https://github.com/WordPress/gutenberg/blame/update/cast-string-parent-to-array/packages/block-library/src/index.native.js#L136) tell otherwise), but specially because the PR worked in my tests.

> Additionally can we simplify the checks here to explicitly check what we want?

What we want to do is "if the parent is a string, normalize it to an array". In my view, the current code is actually more explicit, simpler (why to check both "is it not an array" and "it is a literal string?"), and more comprehensive (String objects).

Reply:
So, we actually have to use settings, just not in `registerBlockType` but in `processBlockType`. This is the data flow for block registration:

- `registerBlockType( nameOrMetadata, settings )`
  - `addBootstrappedBlockType` => if metadata provided, creates a new tree in the store with that data
  - `addUnprocessedBlockType`
    - calls `processBlockType` => takes the existing block metadata & merges it with the settings
      - **merges metadata & settings** into `blockType`
      - dispatch `'blocks.registerBlockType'` filter with `blockType`
      - validates the `blockType` after the filter has been dispatched
  - return `blockType` (settings & metadata consolidated)

I've moved it all to `processBlockType` (including existing validation for the parent that was done too early) in 002ac7cec3337f5f5d56c9f154a3e337b890f3c4.

Reply:
After having looked at this more deeply, there's a situation in which this could be considered a breaking change:

- 3rd party code registers the blockType with parent as a string.
  - Before: parent will be a string in the store.
  - After: parent will be an array in the store.

If the 3rd party code queries the store for the parent _after registration_ to do something with the `parent`, it will no longer be a string. I don't see a way out of this and it seems to me that this PR fixes an important issue (being able to add a child block within itself).

Thoughts?

Reply:
>which this could be considered a breaking change

That's true, but I'm not sure if we should consider a breaking change something that was used in a wrong way and differently than documentation. We had a similar issue with `description` in the past, where some plugins would provide a component and we handled that the same way as you did (with a small difference `deprecated` vs `warning`).

Additionally, it sounds unusual to use `parent` from store and this PR seems good to me regarding this aspect.

Reply:
> We can play it safe and communicate this as a breaking change in dev notes, even though it isn't.

Added the label and [some text](https://github.com/WordPress/gutenberg/pull/66250#issuecomment-2429493919) for the devnote.